### PR TITLE
Split out `TypeRegistry`'s open-coded slab arena into into a reusable type

### PR DESCRIPTION
**Pull request by fitzgen**
First of all, it is just a nice separation of concerns.

Second of all, as I design the GC rooting APIs for Wasmtime's upcoming Wasm GC support, I want this same thing and I'd rather not open code it multiple times.

### CODE REVIEW

File: crates/slab/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    /// A vacant entry.
+    Free {
+        /// A link in the slab's free list, pointing to the next free entry, if
+        /// any.
+        next_free: Option<EntryIndex>,
+    },
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct EntryIndex(NonZeroU32);
```
Comment (elliottt) [inline crates/slab/src/lib.rs:191]: Does this use `NonZeroU32` to ensure that `Option<Id>` takes up four bytes?

**Review by elliottt — APPROVED** [review]
This seems good to me! We could probably use this for the wasi table in wasmtime-wasi as well, as it looks pretty similar in its implementation :)

File: crates/slab/src/lib.rs (hunk 1 above)
Comment (fitzgen) [reply crates/slab/src/lib.rs:191]: Yep, and this also aligns with how `VMSharedTypeIndex` reserves `u32::MAX` for its default/null value.

**Comment by fitzgen** [comment]
> We could probably use this for the `ResourceTable` as well, as it looks pretty similar in its implementation :)

I believe Alex added a generation to that table, so it would need to do the layering described in the docs. Might make sense to add a generic implementation of that layering to this crate if we make `ResourceTable` use this as an implementation detail.

**Pull request merged**